[lua] Add generic logic for Mog Kupons, and data for first 5 kupons

zach2good opened the pull request:
<!-- remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm: -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

## What does this PR do?

Adds the driver code for handling kupon items and KIs, and adds the data for the Dyna Beau + Xarc kupons. Others won't do anything until the data is entered.

## How to test

- Go to dealer moogle in windy: !pos -65.377 -4.188 144.133 240
- Speak to them
- Try to trade garbage to them - nothing happens
- Give yourself Dynamis - Beaucedine kupon: !additem 2745
- Speak to moogle again - nothing
- Trade them the Dynamis - Beaucedine kupon
- Select your chosen gift
- Get the item
- You should have no kupon item or key item, only the prize
- Give yourself Dynamis - Xarc kupon: !additem 2746
- Trade the kupon
- Bail out of the menu
- Get given the dynamis xarc kupon KI
- Talk to the moogle again, be offered the prize for that kupon
- bail out again - nothing happens, you still have the KI
- Fill your inventory
- Talk to moogle, try to get prize, it should fail, you should still have the KI
- Clear your inv
- talk to moogle, get prize, KI should be gone
- You should have 2 bits of prize gear now, no kupon items, no kupon KIs

zach2good committed:
[lua] Add generic logic for Mog Kupons and data for Kupons: A-DBcd and A-DXar

zach2good commented:
It took ages, but I finally figured everything out and this is ready to review :)

TeoTwawki commented on scripts/globals/dealer_moogle.lua:
```diff
@@ -0,0 +1,545 @@
+-- TODO: Where is Kupon AW-UWIII
+
+-- WARNING: These items cannot be customised, and must be in the order
+-- WARNING: they appear in the in-game menus!
```
```
-- WARNING: 
-- These items cannot be customized, 
-- and must be in the order they appear in the in-game menus!
```
_suggested_ revision. not a big deal but my brain had to read twice do to line break placement and double "warning"

claywar commented on scripts/globals/dealer_moogle.lua:
```diff
@@ -0,0 +1,545 @@
+    ]]
+}
+
+-- WARNING: These items cannot be customised!
```
Just a styling thing, but would it be possible to move this table above the item list?  For someone reading top to bottom to understand what the keys in that table are for (and not just linear).  Either that, or a related comment above similar to here where these values are client-defined

zach2good committed:
[lua] Add kupon entries for MOG_KUPON_AW_ABS, MOG_KUPON_AW_PAN and MOG_KUPON_A_LUM

Includes some tidying up of dealer_moogle.lua and adding relevant items to global lists

zach2good commented:
Review items done, added 3 more kupons and tidied a couple of other bits
